component/sqs/component.go

```diff
@@ -0,0 +1,332 @@
+			QueueUrl:            &c.queueURL,
+			MaxNumberOfMessages: c.maxMessages,
+			WaitTimeSeconds:     c.pollWaitSeconds,
+			VisibilityTimeout:   c.visibilityTimeout,
+			AttributeNames: aws.StringSlice([]string{
+				sqsAttributeSentTimestamp,
+			}),
+			MessageAttributeNames: aws.StringSlice([]string{
+				sqsMessageAttributeAll,
+			}),
+		})
+		if err != nil {
+			logger.Errorf("failed to receive messages: %v, sleeping for %v", err, c.retryWait)
+			time.Sleep(c.retryWait)
+			retries--
+			if retries > 0 {
+				continue
+			}
+			chErr <- err
+			return
+		}
+		retries = c.retries
+
+		if ctx.Err() != nil {
+			return
+		}
+
+		logger.Debugf("Consume: received %d messages", len(output.Messages))
+		messageCountInc(c.queueName, fetchedMessageState, len(output.Messages))
+
+		if len(output.Messages) == 0 {
+			continue
+		}
+
+		btc := batch{
+			ctx:       ctx,
+			queueName: c.queueName,
+			queueURL:  c.queueURL,
+			sqsAPI:    c.sqsAPI,
+			messages:  make([]Message, 0, len(output.Messages)),
+		}
+
+		for _, msg := range output.Messages {
+			observerMessageAge(c.queueName, msg.Attributes)
+
+			corID := getCorrelationID(msg.MessageAttributes)
+
+			sp, ctxCh := trace.ConsumerSpan(ctx, trace.ComponentOpName(consumerComponent, c.queueName),
+				consumerComponent, corID, mapHeader(msg.MessageAttributes))
+
+			ctxCh = correlation.ContextWithID(ctxCh, corID)
+			logger := log.Sub(map[string]interface{}{correlation.ID: corID})
+			ctxCh = log.WithContext(ctxCh, logger)
+
+			btc.messages = append(btc.messages, message{
+				ctx:       ctxCh,
+				queueName: c.queueName,
+				queueURL:  c.queueURL,
+				queue:     c.sqsAPI,
+				msg:       msg,
+				span:      sp,
+			})
+		}
+
+		c.proc(ctx, btc)
+	}
+}
+
+func (c *Component) report(ctx context.Context, sqsAPI sqsiface.SQSAPI, queueURL string) error {
+	log.Debugf("retrieve stats for SQS %s", c.queueName)
+	rsp, err := sqsAPI.GetQueueAttributesWithContext(ctx, &sqs.GetQueueAttributesInput{
+		AttributeNames: []*string{
+			aws.String(sqsAttributeApproximateNumberOfMessages),
+			aws.String(sqsAttributeApproximateNumberOfMessagesDelayed),
+			aws.String(sqsAttributeApproximateNumberOfMessagesNotVisible),
+		},
+		QueueUrl: aws.String(queueURL),
+	})
+	if err != nil {
+		return err
+	}
+
+	size, err := getAttributeFloat64(rsp.Attributes, sqsAttributeApproximateNumberOfMessages)
+	if err != nil {
+		return err
+	}
+	queueSize.WithLabelValues("available").Set(size)
+
+	size, err = getAttributeFloat64(rsp.Attributes, sqsAttributeApproximateNumberOfMessagesDelayed)
+	if err != nil {
+		return err
+	}
+	queueSize.WithLabelValues("delayed").Set(size)
+
+	size, err = getAttributeFloat64(rsp.Attributes, sqsAttributeApproximateNumberOfMessagesNotVisible)
+	if err != nil {
+		return err
+	}
+	queueSize.WithLabelValues("invisible").Set(size)
+	return nil
+}
+
+func getAttributeFloat64(attr map[string]*string, key string) (float64, error) {
+	valueString := attr[key]
```

Comment:
would it make sense to use something like 
```
if valueString,ok := attr[key]; ok {
 ... 
}
```

Reply:
actually the value of the map is a *string. so it is nil if the value is not in the map or if the value exists in the map but is actually nil. if i follow your recommendation i have to check again if the valueString is nil.